File: scalaDemo/src/main/scala/DotDemo.scala

object DotDemo extends App {

  // val is default
  case class Dot(x: Int, y: Int)

  implicit class DotConverters(dot: Dot) {
    def convertToJson(): String =
      s"""{"x"=${dot.x}, "y"=${dot.y}}"""
    def convertToXml(): String =
      s"""<dot>
            <x>${dot.x}</x>
            <y>${dot.y}</y>
      </dot>"""
  }

  val dot = Dot(1, 2)
  println("-------- JSON -----------")
  println(dot.convertToJson())
  println("-------- XML  -----------")
  println(dot.convertToXml())
}

File: kotlinDemo/src/main/kotlin/DotDemo.kt
data class Dot (val x: Int, val y: Int)

// неявно получаем ссылку на объект
fun Dot.convertToJson(): String =
        "{\"x\"=$x, \"y\"=$y}"

fun Dot.convertToXml(): String =
        """<dot>
            <x>$x</x>
            <y>$y</y>
        </dot>"""


fun main() {
    val dot = Dot(1, 2)
    println("-------- JSON -----------")
    println(dot.convertToJson())
    println("-------- XML  -----------")
    println(dot.convertToXml())
}

